analytical_engine/apps/simple_path/all_simple_paths.h

```diff
@@ -0,0 +1,190 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef ANALYTICAL_ENGINE_APPS_SIMPLE_PATH_ALL_SIMPLE_PATHS_H_
+#define ANALYTICAL_ENGINE_APPS_SIMPLE_PATH_ALL_SIMPLE_PATHS_H_
+
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "simple_path/all_simple_paths_context.h"
+
+#include "core/app/app_base.h"
+#include "core/worker/default_worker.h"
+
+namespace gs {
+
+template <typename FRAG_T>
+class AllSimplePaths : public AppBase<FRAG_T, AllSimplePathsContext<FRAG_T>>,
+                       public grape::Communicator {
+ public:
+  INSTALL_DEFAULT_WORKER(AllSimplePaths<FRAG_T>, AllSimplePathsContext<FRAG_T>,
+                         FRAG_T)
+  static constexpr grape::MessageStrategy message_strategy =
+      grape::MessageStrategy::kAlongOutgoingEdgeToOuterVertex;
+  static constexpr grape::LoadStrategy load_strategy =
+      grape::LoadStrategy::kBothOutIn;
+  using vertex_t = typename fragment_t::vertex_t;
+  using vid_t = typename fragment_t::vid_t;
+
+  void PEval(const fragment_t& frag, context_t& ctx,
+             message_manager_t& messages) {
+    vertex_t source;
+    vid_t source_gid;
+
+    struct timeval t;
+    gettimeofday(&t, 0);
+    ctx.exec_time -= static_cast<double>(t.tv_sec) +
+                     static_cast<double>(t.tv_usec) / 1000000;
+
+    frag.Oid2Gid(ctx.source_id, source_gid);
+    ctx.soucre_fid = (fid_t) (source_gid >> ctx.fid_offset);
+    bool native_source = frag.GetInnerVertex(ctx.source_id, source);
+    if (native_source) {
+      ctx.source_flag = true;
+      int total_vertex_num = (int) frag.GetTotalVerticesNum();
+      VLOG(0) << "before init map: " << std::endl;
+      ctx.edge_map.resize(total_vertex_num);
+      // for(long unsigned int i=0;i<ctx.edge_map.size();i++){
+      //   ctx.edge_map[i].resize(total_vertex_num,false);
+      // }
+      VLOG(0) << "after init map: " << std::endl;
+    } else {
+      vid_t in_num = frag.GetInnerVerticesNum();
+      fid_t fid = frag.fid();
+      std::tuple<std::pair<vid_t, vid_t>, bool, bool> msg =
+          std::make_tuple(std::make_pair((vid_t) fid, in_num), false, true);
+      messages.SendToFragment(ctx.soucre_fid, msg);
+    }
+
+    gettimeofday(&t, 0);
+    ctx.exec_time += static_cast<double>(t.tv_sec) +
+                     static_cast<double>(t.tv_usec) / 1000000;
+
+    messages.ForceContinue();
+  }
+
+  void IncEval(const fragment_t& frag, context_t& ctx,
+               message_manager_t& messages) {
+    int init_counter = 0;
+    std::tuple<std::pair<vid_t, vid_t>, bool, bool> msg;
+
+    struct timeval t;
+    gettimeofday(&t, 0);
+    ctx.exec_time -= static_cast<double>(t.tv_sec) +
+                     static_cast<double>(t.tv_usec) / 1000000;
+
+    while (messages.GetMessage(msg)) {
+      vid_t gid = std::get<0>(msg).first;
+      vid_t msg2 = std::get<0>(msg).second;
```

Comment:
frag_vnum

Reply:
msg2 not all cases are frag_ vnum

Reply:
you should add a comment to specify what case is frag_vnum, what case  is other ?

Reply:
not `msg2`, a better self-described name?

Reply:
add comment like:
      // if msg[2] == true, msg[0].first is frag_id and msg[0].second is frag's
      // inner_vertex_num.
      // Else if msg[1] == false, means the message is the
      // vertices waiting to be accessed, msg[0].first is frag's gid and
      // msg[0].second is vertex's depth.
      // Else mgs[1] == true, means the message is send to source frag to update
      // edge_map. The msg[0].first is from's gid and msg[0].second is to's gid.